datalad/distribution/tests/test_publish.py

```diff
@@ -220,7 +220,7 @@ def test_publish_plain_git(origin, src_path, dst_path):
     # amend and change commit msg in order to test for force push:
     source.repo.commit("amended", options=['--amend'])
     # push should be rejected (non-fast-forward):
-    assert_raises(IncompleteResultsError,
+    assert_raises(CommandError,
                   publish, dataset=source, to='target', result_xfm='datasets')
```

Comment:
This change in behavior looks problematic both from a presentation standpoint, as well as a results handling standpoint.  For example, consider this script that sets up a superdataset with two subdatasets, one where the push will fail.

<details>
<summary>script</summary>

```sh
#!/bin/sh

set -eux

cd "$(mktemp -d --tmpdir dl-XXXXXXX)"
datalad create origin
(
    cd origin
    datalad create z-fail
    datalad create a-pass
    datalad save --recursive
    git config receive.denyCurrentBranch false
    git -C a-pass config receive.denyCurrentBranch false
)

datalad install --recursive --source origin loc

(
    cd loc
    git -C z-fail commit --allow-empty -m c-z-fail
    git -C a-pass commit --allow-empty -m c-a-pass
    datalad save --recursive
    datalad --on-failure=continue publish --recursive || :
    git log --graph --decorate --oneline master origin/master
    git -C z-fail log --graph --decorate --oneline master origin/master
    git -C a-pass log --graph --decorate --oneline master origin/master
)
```

</details>

Before (specifically on 7888e047c), `datalad --on-failure=continue publish --recursive` would show this output:

```diff
[...]
+ datalad --on-failure=continue publish --recursive
[INFO] Publishing <Dataset path=/tmp/dl-FqwYUhV/loc/z-fail> to origin 
[ERROR] failed to push to origin: master -> origin/master [remote rejected] (branch is currently checked out); pushed: ['58337eb..641a3f5'] [publish(/tmp/dl-FqwYUhV/loc/z-fail)] 
[INFO] Publishing <Dataset path=/tmp/dl-FqwYUhV/loc/a-pass> to origin 
[INFO] Publishing <Dataset path=/tmp/dl-FqwYUhV/loc> to origin 
publish(error): z-fail (dataset) [failed to push to origin: master -> origin/master [remote rejected] (branch is currently checked out); pushed: ['58337eb..641a3f5']]
publish(ok): a-pass (dataset) [pushed to origin: ['1260c8d..31794ac', 'f1e353f..9d1579c']]
publish(ok): . (dataset) [pushed to origin: ['30562bc..eb62938', '8d86f4f..4a88467']]
action summary:
  publish (error: 1, ok: 2)
+ :
+ git log --graph --decorate --oneline master origin/master
* 4a88467 (HEAD -> master, origin/master, origin/HEAD, synced/master) [DATALAD] Recorded changes
* 8d86f4f [DATALAD] Recorded changes
* 0e32a6d [DATALAD] new dataset
+ git -C z-fail log --graph --decorate --oneline master origin/master
* cace822 (HEAD -> master, synced/master) c-z-fail
* ccb080f (origin/master, origin/HEAD) [DATALAD] new dataset
+ git -C a-pass log --graph --decorate --oneline master origin/master
* 9d1579c (HEAD -> master, origin/master, origin/HEAD, synced/master) c-a-pass
* f1e353f [DATALAD] new dataset
```

It displayed a result error for the failing subdataset, and, due to the `--on-failure=continue`, it published all the subdatasets that it could.

Now it dumps the command error and doesn't honor `--on-failure=continue`:

```diff
[...]
+ datalad --on-failure=continue publish --recursive
[INFO] Publishing <Dataset path=/tmp/dl-yQVMqHj/loc/z-fail> to origin 
[INFO] Start fetching remotes for <AnnexRepo path=/tmp/dl-yQVMqHj/loc/z-fail (<class 'datalad.support.annexrepo.AnnexRepo'>)> 
[... 36 lines ...]
[INFO] Finished writing objects 
[INFO] Finished pushing remotes for <AnnexRepo path=/tmp/dl-yQVMqHj/loc/z-fail (<class 'datalad.support.annexrepo.AnnexRepo'>)> 
CommandError: command '['git', 'push', '--progress', '--porcelain', 'origin', 'master', 'git-annex']' failed with exitcode 1
Failed to run ['git', 'push', '--progress', '--porcelain', 'origin', 'master', 'git-annex'] at '/tmp/dl-yQVMqHj/loc/z-fail'.
stdout=To /tmp/dl-yQVMqHj/origin/z-fail
 	refs/heads/git-annex:refs/heads/git-annex	1cfffab..2657d54
!	refs/heads/master:refs/heads/master	[remote rejected] (branch is currently checked out)
[... 18 lines ...]
To /tmp/dl-yQVMqHj/origin/z-fail
 	refs/heads/git-annex:refs/heads/git-annex	1cfffab..2657d54
!	refs/heads/master:refs/heads/master	[remote rejected] (branch is currently checked out)
[... 16 lines ...]
error: failed to push some refs to '/tmp/dl-yQVMqHj/origin/z-fail'
+ :
+ git log --graph --decorate --oneline master origin/master
* a8bbb28 (HEAD -> master) [DATALAD] Recorded changes
* c016fc6 (origin/master, origin/HEAD) [DATALAD] Recorded changes
* cda0d7e [DATALAD] new dataset
+ git -C z-fail log --graph --decorate --oneline master origin/master
* f560fa2 (HEAD -> master, synced/master) c-z-fail
* 7f41170 (origin/master, origin/HEAD) [DATALAD] new dataset
+ git -C a-pass log --graph --decorate --oneline master origin/master
* 7f12cdd (HEAD -> master) c-a-pass
* ab068e5 (origin/master, origin/HEAD) [DATALAD] new dataset
```

Reply:
Good point. I need to reconsider the error reporting. Thx!

Reply:
Done, and pushed.

Reply:
The outcome of the script I posted above is the same after 57cc112096 (BF: Restore previous behavior on push errors, 2020-02-13).  I haven't really looked into it, but I'd guess that is because a CommandError is raised unless a specific error is matched.  When I initially raised the issue, I had assumed the `publish._push` would need to catch the CommandError's in order to retain the previous behavior.